v2/googlecloud-to-neo4j/src/main/java/com/google/cloud/teleport/v2/neo4j/model/helpers/MappingListener.java

```diff
@@ -0,0 +1,505 @@
+import org.neo4j.importer.v1.targets.NodeRangeIndex;
+import org.neo4j.importer.v1.targets.NodeUniqueConstraint;
+import org.neo4j.importer.v1.targets.PropertyMapping;
+import org.neo4j.importer.v1.targets.RelationshipExistenceConstraint;
+import org.neo4j.importer.v1.targets.RelationshipKeyConstraint;
+import org.neo4j.importer.v1.targets.RelationshipRangeIndex;
+import org.neo4j.importer.v1.targets.RelationshipUniqueConstraint;
+
+interface MappingListener {
+  default void enterArray() {}
+
+  default void exitArray() {}
+
+  default void enterObject() {}
+
+  default void enterObjectEntry(String field, String property) {}
+
+  default void exitObject() {}
+
+  default void enterString(String value) {}
+
+  default void exitString() {}
+}
+
+class MappingListeners implements MappingListener {
+  private final Collection<MappingListener> listeners;
+
+  private MappingListeners(Collection<MappingListener> listeners) {
+    this.listeners = listeners;
+  }
+
+  public static MappingListeners of(MappingListener first, MappingListener... rest) {
+    var listeners = new ArrayList<MappingListener>();
+    listeners.add(first);
+    Collections.addAll(listeners, rest);
+    return new MappingListeners(listeners);
+  }
+
+  @Override
+  public void enterArray() {
+    listeners.forEach(MappingListener::enterArray);
+  }
+
+  @Override
+  public void exitArray() {
+    listeners.forEach(MappingListener::exitArray);
+  }
+
+  @Override
+  public void enterObject() {
+    listeners.forEach(MappingListener::enterObject);
+  }
+
+  @Override
+  public void enterObjectEntry(String field, String property) {
+    listeners.forEach(listener -> listener.enterObjectEntry(field, property));
+  }
+
+  @Override
+  public void exitObject() {
+    listeners.forEach(MappingListener::exitObject);
+  }
+
+  @Override
+  public void enterString(String value) {
+    listeners.forEach(listener -> listener.enterString(value));
+  }
+
+  @Override
+  public void exitString() {
+    listeners.forEach(MappingListener::exitString);
+  }
+}
+
+class PropertyMappingListener implements MappingListener {
+  private final Map<String, PropertyMapping> mappingsByProperty;
+  private final BiFunction<String, String, PropertyMapping> newMapping;
+
+  PropertyMappingListener(
+      Map<String, PropertyMapping> mappingsByProperty,
+      BiFunction<String, String, PropertyMapping> newMapping) {
+
+    this.mappingsByProperty = mappingsByProperty;
+    this.newMapping = newMapping;
+  }
+
+  @Override
+  public void enterObjectEntry(String field, String property) {
+    mappingsByProperty.put(field, newMapping.apply(field, property));
+  }
+
+  @Override
+  public void enterString(String value) {
+    mappingsByProperty.put(value, newMapping.apply(value, value));
+  }
+}
+
+/**
+ * This listener groups all properties together under the same key constraint Because of the legacy
+ * spec ambiguity, there as many created constraints as node target labels.
```

Comment:
nit: maybe a bit of rephrasing would help?

Reply:
new attempt in c63df2c0946288da8201d3d4cad05ba4da969b86
wdyt?